### DISCUSSION TITLE
DataCoord memory keep increasing

### DISCUSSION THREAD

**Discussion by byeongjokim** [milvus-io/milvus#28096] [Q&A and General discussion] [answered]
datacoord memory keep increasing and pod died

Could you tell me why this issue occurs and how to prevent it?

<img width="1843" alt="image" src="https://github.com/milvus-io/milvus/assets/17046036/84aedcd4-05ae-4eb9-86f8-4ca0fce23a5b">

- milvus: 2.2.13

**Comment by yhmo** [comment]
How much memory is assigned to the datacoord?

**Reply by byeongjokim** [reply]
@yhmo 
it was 1Gi.

**Reply by yhmo** [reply, marked as answer]
Assign more memory(2GB) to this datacoord, the memory usage should less than a certain value. If it still OOM with 2GB memory, then it could be a memory leak bug. So far as I know, we didn't get any memory leak issues for datacoord.

**Comment by byeongjokim** [comment]
@yhmo 
The memory usage of datacoord(memory: 2GB) continues to grow.
Could you check that there is no problem with memory leak in milvus v2.2.13?

<img width="828" alt="image" src="https://github.com/milvus-io/milvus/assets/17046036/edbf008e-3d79-4b81-86c3-e3831c4270a1">

**Reply by xiaofan-luan** [reply]
Could you offer us a pprof?
Would be even better if it's two pprof that could compare

**Reply by xiaofan-luan** [reply]
@yanliang567 
please also retry to watch in house, seems it is long running issue and leak small amount of data each time

**Reply by byeongjokim** [reply]
@xiaofan-luan 

could you check this pprof results?
this is first time use pprof..
i run command
```
# in datacoord
curl http://127.0.0.1:9091/debug/pprof/heap > heap_snapshot.heap

# cp heap_snapshot.heap to local

$ go tool pprof -png heap_snapshot.heap
```


[heap_snapshot.heap.zip](https://github.com/milvus-io/milvus/files/13316737/heap_snapshot.heap.zip)

![profile001](https://github.com/milvus-io/milvus/assets/17046036/39b92dc6-420a-4c84-83fa-ca0273dd69f3)

**Reply by xiaofan-luan** [reply]
Seems like it's not leak.
Did you insert more message?
Seems that we cache segemnt info into datacoord into memory and it is large.

How many collection, segments, entities are there in your cluster?

**Reply by xiaofan-luan** [reply]
SaveBinlog path can be optimized but so far I think the memory increase is due to data increase

**Reply by xiaofan-luan** [reply]
and most likely it is the log path takes memory
<img width="1355" alt="image" src="https://github.com/milvus-io/milvus/assets/83447078/ed8669c9-9110-4bc3-9d2e-d5505119ec58">

**Reply by byeongjokim** [reply]
@xiaofan-luan 
3 collections
276,029,888 entities
1.48k flushed segment

**Reply by byeongjokim** [reply]
@xiaofan-luan 

The current peak memory records up to 98%.
Is there a config to prevent oom errors?

For example, if I adjust the [maxBinlogFileNumber](https://github.com/milvus-io/milvus/blob/267c67dfee1cefc8acfd2f9dbf2e188d7df317ad/configs/milvus.yaml#L367C5-L367C24) low, can the segment be sealed quickly and prevent OOM?

**Reply by xiaofan-luan** [reply]
THis could definitely improved 
We will test in house on this situation also.
so far our suggestion to give more memory for datacoord, for instance - 4g

**Reply by byeongjokim** [reply]
@xiaofan-luan 

There is a misunderstanding.
The current datacoord memory is 4Gi...

After the test, please share the results asap.
thank you.

**Reply by xiaofan-luan** [reply]
From the flame graph you shared, we only saw 1GB memory usage?

**Reply by xiaofan-luan** [reply]
when you say 4GB, did it mean only datacoord?
how large is your pod limit and request?

**Reply by byeongjokim** [reply]
@xiaofan-luan 

limit: 4GB

There is the other question..

i set [active-standby](https://milvus.io/docs/coordinator_ha.md) for ha (replicas=2).
However there is 3 cases.
Assume that, there is A, B two datacoord pods.

1. An ``active`` A pod died with oom error caused by above memory problem. -> B pod is still in ``standby`` mode.
2. I delete ``active`` A pod using kubectl command (``kubectl delete pod --force`` ). -> B pod is changed to ``active`` mode.
3. I generate oom error on ``active`` A pod using stress-ng command (``stress-ng --brk 2 --stack 2 --bigheap 2``). -> B pod is changed to ``active`` mode.

the cases 2, 3 mean the active-standby works,
however the case 1 mean the active-standby doesnot work.

are there any differences between three cases?

in our guess...
case 1: The A pod died and again lived, in between the period of liveness metric for active-standby.
